### ISSUE TITLE
Two Fer and Error Handling: handling "blank" and leading/trailing whitespaces

### ISSUE THREAD

**Issue by braoult** [exercism/bash#407]
I believe these exercises should be amended to have "blank" strings made equivalent to empty, and leading and trailing blanks being removed for display. For example:

```
$ two_fer.sh " "
One for you, one for me.
$ ./two_fer.sh "   Jean Jacques  "
One for Jean Jacques, one for me.
```
and
```
$ error_handling.sh ""
Usage: ./error_handling <greetee>
$ error_handling.sh "  "
Usage: ./error_handling <greetee>
$ error_handling.sh "   Jean Jacques   "
Hello, Jean Jacques
```

**Comment by IsaacG** [comment]
This exercise focuses on error handling when an argument is missing. (1) I'm not convinced we want to add additional complexity with parsing the arg. (2) I'm not sure modifying the arg would even be considered correct behavior.

**Comment by glennj** [comment]
I sometimes lose track of the purpose of the exercises. 

Error handling is focused on simply identifying the given error condition, in this case the number of parameters. I can see testing if the argument is empty is an error, but extra spaces is probably not. (Although I can now envision some other error conditions that would be worthwhile testing for.)

Two-fer's focus is to provide a default value if the argument is missing or empty. Trimming whitespace is not on point here.

**Comment by braoult** [comment]
I agree, this exercise is not to parse arguments...

But we have to go to test script to understand what is absolutely necessary.
The description being not explicit enough, everybody will have to understand what is needed after failing some tests, instead of being able to solve in one shot.

Either *all* requirements should be explained (change there), either some tests are removed/amended/added. If we need to read the test script to guess the output string is "blabla", we can easily make a script to answer exactly and only the test cases.

I did my best in a few tests, to have useful answers (spaces, etc...), tests failed with my input. I just suggest to accept more answers, when better. Not important, anyway, thanks a lot for your site.

**Comment by glennj** [comment]
That's an interesting perspective. When I approach solving an exercise, I read the README to get an overview of the problem, and then I immediately head to the test file to get the specifications for what API I'm about to implement. I agree that there's an opportunity here to improve the bash-specific portions of the README for this exercise.

**Comment by braoult** [comment]
Why not replacing:
"However, if the name is missing, return the string"
with:
"However, if the name is missing or empty, return the string"
And adding the following test:
```
@test "empty string" {
  [[ $BATS_RUN_SKIPPED == true  ]] || skip
  run bash two_fer.sh ""
  [[ $status -eq 0 ]]
  [[ $output == "One for you, one for me." ]]
}
```
And (forgetting about trimming leading/trailing spaces) if we want to consider the case of "blank only" parameter, we could also add "non blank" to the description, and add the test:
```
@test "whitespaces only" {
  [[ $BATS_RUN_SKIPPED == true  ]] || skip
  run bash two_fer.sh "  "
  [[ $status -eq 0 ]]
  [[ $output == "One for you, one for me." ]]
}
```
I believe this could apply to other exercises. I mean to consider empty (and maybe blank) parameters.

**Comment by IsaacG** [comment]
I would argue that empty strings are still valid arguments and should be used like any other argument. I encourage people to test the number of arguments and not the contents of them. If the user decides to pass "" as an argument, then the script should honor that choice and not second guess or restrict how the user can use it.

**Comment by braoult** [comment]
Obviously, if the exercise explanation is not changed. I was suggesting to change the explanation and the [empty string] test. Both, or none.
I prefer both, but you will have the last word. We discussed enough about it, why not closing this issue if everybody agrees that "One for , one for me" is a correct output ?
I would add (maybe wrongly) that any exercise should have a user goal. Like to output a sentence here. If not, the test could be have been to echo "Argument is [<arg>]", instead of "One for <arg>, one for me". In this case, I would not even have opened this issue, as it would have been clear that empty string would be acceptable.

**Comment by guygastineau** [comment]
What do people want to do about this?  Should we customize the two-fer exercise for the bash track?

**Comment by kotp** [comment]
I think that "One for , one for me." output given an empty string as an argument, but the test could be added to exhibit this explicitly to answer that question.  It may be useful to bring this up for all the tracks as well, as it isn't isolated to only Bash.

**Comment by guygastineau** [comment]
This has been been brought up before for all of the tracks.  I think `bash` is one of the only tracks where `""` is handled like invalid input.  Of course (ignoring the semantics of the exercise) in Bash typically if our script is expecting an argument, then that argument won't be `""`.  If it is, then something probably went wrong.

**Comment by guygastineau** [comment]
So it appears like we could accept either way as valid, and that we would probably benefit from explicitly testing for the case in which `""` is passed as an argument.  For other languages this is a non-issue since they typically know if an argument is present or not.

In bash it is more natural to use `${1:-some default}`.  This will fire even if an empty argument is supplied.  There is a slightly different syntax, that only supplies the default when the first argument is literally unset.  I think the fact that Bash blurs the lines between empty string and unset parameter makes it more difficult to decide what is wright.  Ultimately, I don't care which way we go.  So long as it is consistent and our mentoring notes and readme make the objective clear I am fine with the outcome.

For the record though, in a shell script expecting an argument if I pass an empty string I want a default case to fire 99% of the time.

**Comment by kotp** [comment]
Other languages, I do end up talking about passing in an empty string explicitly, sometimes it isn't "input" from the keyboard, sometimes it may be coming from other sources.  So the idea of what is "acceptable input" including the idea of "nothing" and what to do when we get it is definitely missing from the specifications here... but I don't think that is a bad thing, it happens to be a good conversation and exploration point.

**Comment by guygastineau** [comment]
### TL;DR
So I think that the first way we did this in bash was the right way to do it.  I think it is the right way to do it, because:
1. We can show ourselves how it is theoretically tantamount to abstracting the possibility of `something not-being` (`Maybe` or any option type), and then lifting the empty case of the monoid string into that abstraction.
2. There is some convention and precedent in shell environments to treat unset and empty environment variables similarly, and these are also often treated similarly to missing or empty function arguments.

-----

@kotp This just got me thinking.  The last time this came up in a broader part of exercism (in the problem-specifications I think) there were some really passionate arguments about what a name could be.  Some argued passionately that one could have no name, and how we can't devalue a lack of name being a name.

Coming at it from the other side, I now think of strings as a `Monoid`.  Regardless of the underlying implementation we are left with some characters or an empty/zero case when it is isolated in the abstract. There is also a binary operatotion `append` that is associative and effectively ignores the empty case.  Thus strings are abstractly monoidal.  Just like natural numbers make a monoid with `+` and another one with `*` (though the empty case changes from `0` to `1`).

Regarding the idea of no input then someone could easily make the case for `""` being the no input.  After all (if `++` appends strings then) `"" ++ "this is " ++ "" ++ "a test"` is the same as `"this is " ++ "a test"`. DISCLAIMER: This seems true for shells, but not likely for any other languages.  Largely due to shell splitting and expansion we have a blurred line between input/arguments and strings.  For example:
``` bash
count_args () { echo $# }
count_args "this is a test"
# => 1
count_args this is a test
# => 4
# but also
[[ $(echo "this is a test") == $(echo this is a test) ]] && echo true ||
  echo false
# => true
```

At the same time, in the `real world` (messy realm of IO I suppose) things can happen where no argument is given at all.  This is especially true for a language like Bash which is constantly wrapped up in IO.

I think these are some of the reasons why no one has bothered to port the two-fer exercise to the Haskell track.  If we leverage the natural monoid `String` then the solution would be:
``` haskell
twoFer :: String -> String
twoFer ""   = twoFer "you"
twoFer name = "One for " ++ name ++ ", one for me."
```
If people want to respect the possibility of an empty name, then it must be wrapped up in another Monoid:
``` haskell
twoFer :: Maybe String -> String
twoFer Nothing     = twoFer "you"
twoFer (Just name) = "One for " ++ name ++ ", one for me."
```

What this probably makes the most clear is that `${1-you}` functions just like the `Maybe twoFer` if we associate `Maybe` with the potential infelicity introduced by the nature of IO in bash scripts (namely by a script requiring arguments).

`${1:-you}`, however, functions exactly like neither of these Haskell examples.  Instead it manages to equate `Nothing` with the emptiness of an empty string.  In effect it has bound the empty member of the String Monoid with `Nothing` as an abstraction over the potential for nothing to happen (for no arguments to be present).

I would wager that the more type safe a language becomes the more ludicrous this sounds initially, but it is far more common in Bash than `${1:-default}`.  Inspecting this further I propose the following classification of what is happening:
    IO can't be a monoid, because we (at least I) can't establish the associativity rules:
        (a ++ b) ++ c == a ++ (b ++ c)
    But considering a sequence of IO actions we can identify an empty case.  A case in which nothing happens.
Often when IO fails to happen a program has to block or fail.  Hopefully there is a timeout that can be set somewhere.  For the case of command line arguments to shell scripts (and to functions) we have some very convenient ways to handle just the presence/absence of an argument or to equate that with the emptiness of a string as well.

It occurs to me that I think this is very cool, but perhaps I am very far off of the point :smiling_imp: 

-----

I know that was long, but it got me to consider why this behavior exists in Bash, and why it exists in other languages.  I think for a lot of other languages it is used to provide defaults.  For example, if we want to count the string `"cat"` in a list in common lisp we could try:
``` lisp
(count "cat" '("cat" "bat" "tab" "cat"))
```
Interestingly this will return nothing.  `count` has an optional argument `test` which is a binary predicate to use for testing equality.  The default is `eq` which compares memory addresses for boxed values.  Thus, the correct way to get the answer 2 is:
``` lisp
(count "cat" '("cat" "bat "tab" "cat") :test #'equal)
```
In python the `two-fer` exercise can be un-idiomatically satisfied with:
``` python
two_fer = lambda name = 'you': f'One for {name}, one for me.'
```
Once again we are seeing a mechanism that mostly promotes the use of defaults when arguments are not passed.  This is evident when trying to get all non alphanumeric characters in a string converted to space so you can have a clean split to get an array:
``` python
re.sub('[^0-9a-z]', ' ', text, flags = re.IGNORECASE).split()
```
Obviously the defaults deliver case sensitive matching in that case.

-----

So, judging from the last examples there I see examples of sensible defaults being enacted when they don't need to be overridden.  Like the lack of an argument is purposeful.  That almost wants to contradict my preference for understanding this problem in Shell.

Analyzing when I use these features in Bash scripts.  I use these features in Bash religiously when defining `die` functions at the top of scripts:
``` bash
die () {
  msg="${1:-something has gone horribly wrong}"
  printf '%s\n' "$msg" && exit 1
}
```
I guess I also use it a lot for setting values that might get set by the command line or for environmental variables that might not be set.

In literally every one of these case I want the `:-` behavior.  For the `die` function I can leverage the default, but mostly I try to pass informative error messages so the `:-` there is a backup.

When used for variables and environment I guess it is more along the lines of setting sane defaults that can be overridden like in the common lisp example (although we might not consider `#'eq` a sane predicate for `#'count`).

Most utilities that reference our environment variables in Unices also state `if X is unset or an empty string`.

If you made it this far, then I am sorry I am so long winded.  Hopefully this was interesting ;)
The TL;DR at the top is a blatant copy of the following conclusion :smiling_imp: 
-----

### TL;DR
So I think that the first way we did this in bash was the right way to do it.  I think it is the right way to do it, because:
1. We can show ourselves how it is theoretically tantamount to abstracting the possibility of `something not-being` (`Maybe` or any option type), and then lifting the empty case of the monoid string into that abstraction.
2. There is some convention and precedent in shell environments to treat unset and empty environment variables similarly, and these are also often treated similarly to missing or empty function arguments.

**Comment by guygastineau** [comment]
Well, I focused on the difference between nothing and `""`, but now I realize that this issue has more to do with whitespace.

**Comment by kotp** [comment]
I usually will not mention the problematic space after the word before the "empty space" for the reason that this is an early exercise.  Otherwise, `"One for, one for me."` would be the desired result, if an empty space is not expected to be an error.

I say we focus on the default argument, that an empty record (in this case an empty string) is allowed, and in fact, any record at all is acceptable.

**Comment by kotp** [comment]
Also, we shouldn't increment the number (not everyone will have access to edit, and perhaps we don't restrict it to only maintainers), but a voting icon... :+1: :-1:  or :eyes: as can be done from the emote button in that message.

**Comment by guygastineau** [comment]
@kotp Thank you for switching the vote to be emoji based.  I hadn't considered how that closes it off to non-maintainers :smile: 

If `two-fer` wasn't already changed to accept empty names I am willing to have that change happen, but that should have its own issue (it is really not so related to the title of this issue (more related to my accidental tangents in this issue)).

I also, think that it should be discussed in its own issue.  There are some implications to consider.  Namely, `:-` is considered the better/safer/more-predictable practice over `-`, but we will be teaching a lot of people to use `-`.  Anyway, I will leave the rest of this for a dedicated issue.

**Comment by kotp** [comment]
If you have not done so yet, @guygastineau you can click on the `...` to the top right of your [comment](https://github.com/exercism/bash/issues/407#issuecomment-570070761) and "reference in new issue".
